testcases/kernel/security/stack_clash/stack_clash.c

```diff
@@ -0,0 +1,165 @@
+ * considered hard to hop above. Code adapted from the Novell's bugzilla [2].
+ *
+ * [1] https://blog.qualys.com/securitylabs/2017/06/19/the-stack-clash
+ * [2] https://bugzilla.novell.com/show_bug.cgi?id=CVE-2017-1000364 */
+
+#include <sys/mman.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <alloca.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include "tst_test.h"
+
+static long PAGE_SIZE;
+static long PAGE_MASK;
+static int STACK_GROWSDOWN;
+
+#define SIGNAL_STACK_SIZE	(1UL<<20)
+#define MEGABYTE		(1<<20)
+#define THRESHOLD		(250 * PAGE_SIZE)
+#define FRAME_SIZE		1024
+
+void recurse(void)
+{
+	volatile char * ptr = alloca(FRAME_SIZE - sizeof(long));
+	*ptr = '\0';
+	recurse();
+}
+
+#define MAPPED_LEN PAGE_SIZE
+static unsigned long mapped_addr;
+void segv_handler(int sig, siginfo_t *info, void *data)
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV)
+		return;
+
+	(void) data;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO, "mmap = [%lx, %lx), addr = %lx, diff = %lx",
+		mapped_addr, mmap_end, fault_addr, diff);
+	if (diff < THRESHOLD)
+		exit(EXIT_FAILURE);
+	else
+		exit(EXIT_SUCCESS);
+}
+
+void do_child(void)
+{
+	int stack_top;
+	unsigned long stack_addr = (unsigned long)&stack_top;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	struct rlimit rlimit;
+
+	SAFE_GETRLIMIT(RLIMIT_STACK, &rlimit);
+	if (rlimit.rlim_cur == RLIM_INFINITY)
+		rlimit.rlim_cur = 16 * MEGABYTE;
+
+	if (STACK_GROWSDOWN)
+		mapped_addr = stack_addr - rlimit.rlim_cur;
+	else
+		mapped_addr = stack_addr + rlimit.rlim_cur;
+
+	mapped_addr &= PAGE_MASK;
+	SAFE_MMAP((void *)mapped_addr, MAPPED_LEN,
+			PROT_READ|PROT_WRITE,
+			MAP_ANON|MAP_PRIVATE|MAP_FIXED, -1, 0);
+	tst_res(TINFO, "Stack top:0x%lx mmap:0x%lx", stack_addr, mapped_addr);
+
+
+	/* Make sure that our SEGV handler will have a stack to run on */
+	signal_stack.ss_sp = SAFE_MALLOC(SIGNAL_STACK_SIZE);
+	signal_stack.ss_size = SIGNAL_STACK_SIZE;
+	signal_stack.ss_flags = 0;
+	if (sigaltstack(&signal_stack, NULL) == -1) {
+		perror("sigaltstack " __FILE__);
+		return;
+	}
+	if (sigaction(SIGSEGV, &segv_sig, NULL) == -1) {
+		perror("sigaction " __FILE__);
+		return;
+	}
+
+	recurse();
+	/* Should never return */
+}
+
+void setup(void)
+{
+	PAGE_SIZE = sysconf(_SC_PAGESIZE);
+	PAGE_MASK = ~(PAGE_SIZE - 1);
+
+	{
+		volatile int *a = alloca(128);
+
+		{
+			volatile int *b = alloca(128);
+
+			STACK_GROWSDOWN = b < a;
+			tst_res(TINFO, "STACK_GROWSDOWN = %d, %p, %p", STACK_GROWSDOWN, b, a);
+		}
+	}
+}
+
+void stack_clash_test(void)
+{
+	int status;
+	pid_t pid;
+
+	pid = SAFE_FORK();
+	if (!pid) {
+		do_child();
+		exit(EXIT_FAILURE);
+		return;
+	}
+
+	SAFE_WAITPID(pid, &status, 0);
+
+	if (WIFEXITED(status)) {
+		if (WEXITSTATUS(status) == 1)
```

Comment:
When using 'tst_brk' as I suggested this should be `WEXITSTATUS(status) != 0` because it might not return 1 or 0.